README.md
dnsblock(1) -- Regex Hosts file
===============================

## MISSION

It becomes harder and harder to control to whom your computer wants to talk. While visiting a web page your computer contacts tracking services or even malware platforms all over the world. This may happen within email programs as well. Cloud services are fast moving targets with various backends and vendors. This line of defense focuses on DNS.

## DESCRIPTION

*dnsblock* is a NSS library that acts like the /etc/hosts file except that it makes use of regular expressions (PCRE) to match the hostname. Create a configuration file ~/.dnsblock and add *dnsblock* to your /etc/nsswitch.conf configuration to enable *dnsblock*.

## INSTALLATION

There is a simple installation helper for Debian/Ubuntu. Just run one of the
following commands to create a installable package.

	$ make debian
	$ make ubuntu

Otherwise, you may use

	$ make
	$ sudo make install

## CONFIGURATION

Add dnsblock to /etc/nsswitch.conf for regular use.

	hosts: files dnsblock dns

Add dnsblock_debug for debugging purposes.

	hosts: files dnsblock_debug dns

## EXAMPLES

/etc/dnsblock

	# Early pass-through for all multicast DNS requests.
	* 			\.local$

	# Block ad servers like adserver.example.net.
	127.0.0.1	^adserver\.

	# Block tracking servers like tracking123.somedomain.com.
	127.0.0.2	^tracking\d*\.

	# Pass-through some TLDs.
	* 			\.nl$
	* 			\.se$
	* 			\.org$

	# Block everything else.
	127.0.0.3	.*

Have a look at *docs/dnsblock.example* for a real life example.

## LOGGING

*dnsblock_debug* logs all DNS requests to your log environment.

	$ tail -f /var/log/messages
	$ journalctl -b -f

## FILES

dnsblock reads /etc/dnsblock and $HOME/.dnsblock in that particular order. Both files get parsed if they exist.


debian/changelog
dnsblock (0.07) stretch; urgency=low

  * New dnsblock_debug module for logging purposes.
  * Do not log PASS and DENY by default.

 -- Aiko Barz <aiko@torrentkino.de>  Sat, 15 Sep 2018 22:57:00 +0100
dnsblock (0.06) stretch; urgency=low

  * Real world example

 -- Aiko Barz <aiko@torrentkino.de>  Sun, 26 Aug 2018 15:22:00 +0100
dnsblock (0.05) wheezy; urgency=low

  * Real world example

 -- Aiko Barz <aiko@torrentkino.de>  Sat, 21 Mar 2015 12:07:00 +0100
dnsblock (0.04) wheezy; urgency=low

  * Use '*' instead of '-' for the pass through rules.

 -- Aiko Barz <aiko@torrentkino.de>  Sat, 21 Mar 2015 08:04:00 +0100

dnsblock (0.03) wheezy; urgency=low

  * Create bypass rules

 -- Aiko Barz <aiko@torrentkino.de>  Thu, 19 Mar 2015 23:21:00 +0100
dnsblock (0.02) wheezy; urgency=low

  * 1st: Load /etc/dnsblock
  * 2nd: Load $HOME/.dnsblock

 -- Aiko Barz <aiko@torrentkino.de>  Fri, 19 Dec 2014 21:27:00 +0100
dnsblock (0.01) wheezy; urgency=low

  * Init

 -- Aiko Barz <aiko@torrentkino.de>  Thu, 18 Dec 2014 11:54:00 +0100


src/nss.c
/*
Copyright 2014 Aiko Barz

This file is part of dnsblock.

dnsblock is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

dnsblock is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with dnsblock.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <nss.h>
#include <netdb.h>
#include <sys/socket.h>
#include <arpa/inet.h>

#include "nss.h"

#ifdef DEBUG
enum nss_status _nss_dnsblock_debug_gethostbyname_r(const char *hostname,
						    struct hostent *host,
						    char *buffer,
						    size_t buflen, int *errnop,
						    int *h_errnop)
{

	return _nss_dnsblock_debug_gethostbyname3_r(hostname, AF_UNSPEC, host,
						    buffer, buflen, errnop,
						    h_errnop, NULL, NULL);
}
#else
enum nss_status _nss_dnsblock_gethostbyname_r(const char *hostname,
					      struct hostent *host,
					      char *buffer,
					      size_t buflen, int *errnop,
					      int *h_errnop)
{

	return _nss_dnsblock_gethostbyname3_r(hostname, AF_UNSPEC, host,
					      buffer, buflen, errnop, h_errnop,
					      NULL, NULL);
}
#endif

#ifdef DEBUG
enum nss_status _nss_dnsblock_debug_gethostbyname2_r(const char *hostname,
						     int af,
						     struct hostent *host,
						     char *buffer,
						     size_t buflen,
						     int *errnop, int *h_errnop)
{
	return _nss_dnsblock_debug_gethostbyname3_r(hostname, af, host,
						    buffer, buflen, errnop,
						    h_errnop, NULL, NULL);
}
#else
enum nss_status _nss_dnsblock_gethostbyname2_r(const char *hostname,
					       int af,
					       struct hostent *host,
					       char *buffer,
					       size_t buflen,
					       int *errnop, int *h_errnop)
{
	return _nss_dnsblock_gethostbyname3_r(hostname, af, host,
					      buffer, buflen, errnop, h_errnop,
					      NULL, NULL);
}
#endif

#ifdef DEBUG
enum nss_status _nss_dnsblock_debug_gethostbyname3_r(const char *hostname,
						     int af,
						     struct hostent *host,
						     char *buffer,
						     size_t buflen, int *errnop,
						     int *h_errnop,
						     int32_t * ttlp,
						     char **canonp)
{
	return _nss_dnsblock_debug_hostent(hostname, strlen(hostname), af, host,
					   buffer, buflen, errnop, h_errnop,
					   ttlp, canonp);
}
#else
enum nss_status _nss_dnsblock_gethostbyname3_r(const char *hostname, int af,
					       struct hostent *host,
					       char *buffer, size_t buflen,
					       int *errnop, int *h_errnop,
					       int32_t * ttlp, char **canonp)
{
	return _nss_dnsblock_hostent(hostname, strlen(hostname), af, host,
				     buffer, buflen, errnop, h_errnop, ttlp,
				     canonp);
}
#endif

#ifdef DEBUG
enum nss_status _nss_dnsblock_debug_gethostbyname4_r(const char *hostname,
						     struct gaih_addrtuple
						     **pat, char *buffer,
						     size_t buflen, int *errnop,
						     int *h_errnop,
						     int32_t * ttlp)
{
	return _nss_dnsblock_debug_gaih_tuple(hostname, strlen(hostname), pat,
					      buffer, buflen, errnop, h_errnop,
					      ttlp);
}
#else
enum nss_status _nss_dnsblock_gethostbyname4_r(const char *hostname,
					       struct gaih_addrtuple **pat,
					       char *buffer, size_t buflen,
					       int *errnop, int *h_errnop,
					       int32_t * ttlp)
{
	return _nss_dnsblock_gaih_tuple(hostname, strlen(hostname), pat,
					buffer, buflen, errnop, h_errnop, ttlp);
}
#endif

#ifdef DEBUG
enum nss_status _nss_dnsblock_debug_hostent(
#else
enum nss_status _nss_dnsblock_hostent(
#endif
					     const char *hostname, int hostsize,
					     int af, struct hostent *host,
					     char *buffer, size_t buflen,
					     int *errnop, int *h_errnop,
					     int32_t * ttlp, char **canonp)
{

	UCHAR address[16];
	char *p_addr = NULL;
	char *p_name = NULL;
	char *p_aliases = NULL;
	char *p_addr_list = NULL;
	char *p_idx = NULL;
	size_t s_total = 0;
	int in_addr_size = 0;
	int result = 0;
	int j = 0;

	/* Check */
	if (!_nss_dnsblock_valid_hostname(hostname, hostsize)) {
		*errnop = ENOENT;
		*h_errnop = HOST_NOT_FOUND;
		return NSS_STATUS_NOTFOUND;
	}

	/* Ask daemon */
	result = _nss_dnsblock_lookup(hostname, address, &af, &in_addr_size);
	if (result == 0) {
		*errnop = ENOENT;
		*h_errnop = HOST_NOT_FOUND;
		return NSS_STATUS_NOTFOUND;
	}

	s_total =
	    (hostsize + 1) * sizeof(char) +
	    sizeof(char *) +
	    result * in_addr_size * sizeof(char) +
	    (result + 1) * sizeof(char *);

	if (buflen < s_total) {
		*errnop = ENOMEM;
		*h_errnop = NO_RECOVERY;
		return NSS_STATUS_TRYAGAIN;
	}

	memset(buffer, '\0', buflen);

	/* Hostname */
	p_name = buffer;
	memcpy(p_name, hostname, hostsize);
	p_idx = p_name + hostsize + 1;

	/* Alias */
	p_aliases = p_idx;
	*(char **)p_aliases = NULL;
	p_idx += sizeof(char *);

	/* Address data */
	p_addr = p_idx;
	memcpy(p_addr, address, result * in_addr_size);
	p_idx += result * in_addr_size;

	/* Address pointer */
	p_addr_list = p_idx;
	p_idx = p_addr;
	for (j = 0; j < result; j++) {
		((char **)p_addr_list)[j] = p_idx;
		p_idx += in_addr_size;
	}
	((char **)p_addr_list)[result] = NULL;

	/* Finish object */
	host->h_name = p_name;
	host->h_aliases = (char **)p_aliases;
	host->h_addrtype = af;
	host->h_length = in_addr_size;
	host->h_addr_list = (char **)p_addr_list;

	if (ttlp != NULL) {
		*ttlp = 0;
	}

	if (canonp != NULL) {
		*canonp = p_name;
	}

	return NSS_STATUS_SUCCESS;
}

#ifdef DEBUG
enum nss_status _nss_dnsblock_debug_gaih_tuple(
#else
enum nss_status _nss_dnsblock_gaih_tuple(
#endif
						const char *hostname,
						int hostsize, struct
						gaih_addrtuple **pat,
						char *buffer, size_t buflen,
						int *errnop, int *h_errnop,
						int32_t * ttlp)
{

	UCHAR address[16];
	char *p_name = NULL;
	char *p_idx = NULL;
	UCHAR *p_addr = NULL;
	struct gaih_addrtuple *p_tuple = NULL;
	struct gaih_addrtuple *p_start = NULL;
	size_t s_total = 0;
	int af = AF_UNSPEC;
	int in_addr_size = 0;
	int result = 0;
	int j = 0;

	/* Check */
	if (!_nss_dnsblock_valid_hostname(hostname, hostsize)) {
		*errnop = ENOENT;
		*h_errnop = HOST_NOT_FOUND;
		return NSS_STATUS_NOTFOUND;
	}

	/* Ask daemon */
	result = _nss_dnsblock_lookup(hostname, address, &af, &in_addr_size);
	if (result == 0) {
		*errnop = ENOENT;
		*h_errnop = HOST_NOT_FOUND;
		return NSS_STATUS_NOTFOUND;
	}

	s_total =
	    (hostsize + 1) * sizeof(char) +
	    result * sizeof(struct gaih_addrtuple);

	if (buflen < s_total) {
		*errnop = ENOMEM;
		*h_errnop = NO_RECOVERY;
		return NSS_STATUS_TRYAGAIN;
	}

	/* Hostname */
	memset(buffer, '\0', buflen);
	p_name = buffer;
	memcpy(p_name, hostname, hostsize);

	/* Object */
	p_idx = p_name + hostsize + 1;
	p_addr = address;
	p_start = (struct gaih_addrtuple *)p_idx;
	for (j = 0; j < result; j++) {
		p_tuple = (struct gaih_addrtuple *)p_idx;
		p_tuple->name = p_name;
		p_tuple->family = af;
		memcpy(p_tuple->addr, p_addr, in_addr_size);
		p_tuple->scopeid = 0;

		/* Linked list */
		if (j == result - 1) {
			p_tuple->next = NULL;
		} else {
			p_tuple->next = (struct gaih_addrtuple *)
			    (p_idx + sizeof(struct gaih_addrtuple));
		}

		p_idx += sizeof(struct gaih_addrtuple);
		p_addr += in_addr_size;
	}

	*pat = p_start;

	if (ttlp != NULL) {
		*ttlp = 0;
	}

	return NSS_STATUS_SUCCESS;
}


src/nss.h
/*
Copyright 2014 Aiko Barz

This file is part of dnsblock.

dnsblock is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

dnsblock is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with dnsblock.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef NSS_H
#define NSS_H

#include "config.h"
#include "main.h"

#define _public_ __attribute__( ( visibility( "default")))
#define _hidden_ __attribute__( ( visibility( "hidden")))

#ifdef DEBUG
enum nss_status _nss_dnsblock_debug_gethostbyname_r(
#else
enum nss_status _nss_dnsblock_gethostbyname_r(
#endif
						     const char *hostname,
						     struct hostent *host,
						     char *buffer,
						     size_t buflen, int *errnop,
						     int *h_errnop) _public_;

#ifdef DEBUG
enum nss_status _nss_dnsblock_debug_gethostbyname2_r(
#else
enum nss_status _nss_dnsblock_gethostbyname2_r(
#endif
						      const char *hostname,
						      int af,
						      struct hostent *host,
						      char *buffer,
						      size_t buflen,
						      int *errnop,
						      int *h_errnop) _public_;

#ifdef DEBUG
enum nss_status _nss_dnsblock_debug_gethostbyname3_r(
#else
enum nss_status _nss_dnsblock_gethostbyname3_r(
#endif
						      const char *hostname,
						      int af,
						      struct hostent *host,
						      char *buffer,
						      size_t buflen,
						      int *errnop,
						      int *h_errnop,
						      int32_t * ttlp,
						      char **canonp) _public_;

#ifdef DEBUG
enum nss_status _nss_dnsblock_debug_gethostbyname4_r(
#else
enum nss_status _nss_dnsblock_gethostbyname4_r(
#endif
						      const char *hostname, struct gaih_addrtuple
						      **pat, char *buffer,
						      size_t buflen,
						      int *errnop,
						      int *h_errnop,
						      int32_t * ttlp) _public_;

#ifdef DEBUG
enum nss_status _nss_dnsblock_debug_hostent(
#else
enum nss_status _nss_dnsblock_hostent(
#endif
					     const char *hostname, int hostsize,
					     int af, struct hostent *host,
					     char *buffer, size_t buflen,
					     int *errnop, int *h_errnop,
					     int32_t * ttlp, char **canonp);

#ifdef DEBUG
enum nss_status _nss_dnsblock_debug_gaih_tuple(
#else
enum nss_status _nss_dnsblock_gaih_tuple(
#endif
						const char *hostname,
						int hostsize, struct
						gaih_addrtuple **pat,
						char *buffer, size_t buflen,
						int *errnop, int *h_errnop,
						int32_t * ttlp);

#endif				/* NSS_H */


src/main.c
/*
Copyright 2014 Aiko Barz

This file is part of dnsblock.

dnsblock is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

dnsblock is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with dnsblock.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <string.h>
#include <stdarg.h>
#include <arpa/inet.h>
#include <syslog.h>
#include <pcre.h>

#include "main.h"

int _nss_dnsblock_lookup(const char *hostname, UCHAR * address,
			 int *af, int *address_size)
{
	int result = 0;
	char filename[256];

	/* 1st: Load global file */
	result = _nss_dnsblock_load("/etc/dnsblock", hostname, address,
				    af, address_size);

	if (result == 1) {
		return 1;
	}

	/* 2nd: Load local file */
	if (getenv("HOME") == NULL) {
		return 0;
	}

	snprintf(filename, 256, "%s/.dnsblock", getenv("HOME"));

	result = _nss_dnsblock_load(filename, hostname, address,
				    af, address_size);

	if (result == 1) {
		return 1;
	}

	return 0;
}

int _nss_dnsblock_load(const char *filename, const char *hostname,
		       UCHAR * address, int *af, int *address_size)
{
	FILE *fp = NULL;
	char str[BUF_SIZE];
	char regex[256];
	char target[256];
	int result = 0, match = 0, is_ip = 0;
	UCHAR buf[sizeof(struct in6_addr)];

	fp = fopen(filename, "r");
	if (fp == NULL) {
		return 0;
	}

	while (fgets(str, BUF_SIZE, fp) != NULL) {
		/* Too big fo regex/target */
		if (strlen(str) >= BUF_SIZE) {
			continue;
		}

		switch (str[0]) {
		case '#':
		case ' ':
		case '\t':
		case '\n':
		case '\0':
			continue;
			break;
		}

		/* Parse /etc/dnsblock */
		sscanf(str, "%255s%*[ \t]%255s", target, regex);

		/* Match requested hostname against my regular expressions */
		match = _nss_dnsblock_pcre_match(hostname, regex);

		/* No match: Go on */
		if (match == 0) {
			continue;
		}

		/* Whitelist: Early passthrough */
		if (*target == '*') {
			result = 1;
			break;
		}

		/* Try to convert the IP address to a binary representation. Start with
		   IPv6. Try IPv4 later. */
		is_ip = inet_pton(AF_INET6, target, buf);
		if (is_ip == 1) {
			*af = AF_INET6;
		} else {
			is_ip = inet_pton(AF_INET, target, buf);
			if (is_ip == 1) {
				*af = AF_INET;
			}
		}

		if (is_ip == 1) {
			*address_size = (*af == AF_INET6) ? 16 : 4;
			memcpy(address, buf, *address_size);
			result = 2;
			break;
		} else {
			_nss_dnsblock_syslog("ERROR: \"%s\"", target);
			continue;
		}
	}
	fclose(fp);

#ifdef DEBUG
	/* Some debugging */
	switch (result) {
	case 0:
		_nss_dnsblock_syslog("PASS: %s", hostname);
		break;
	case 1:
		_nss_dnsblock_syslog("PASS: %s", hostname);
		break;
	default:
		_nss_dnsblock_syslog("DENY: %s -> %s", hostname, target);
	}
#endif

	/* Return 0 or 1 only */
	return (result <= 1) ? 0 : 1;
}

int _nss_dnsblock_pcre_match(const char *subject, const char *pattern)
{
	pcre *re = NULL;
	const char *error = NULL;
	int erroffset = 0;
	int ovector[OVECCOUNT];
	int size = 0;
	int rc = 0;

	size = strlen(subject);

	re = pcre_compile(pattern,	/* the pattern */
			  0,	/* default options */
			  &error,	/* for error message */
			  &erroffset,	/* for error offset */
			  NULL);	/* use default character tables */

	if (re == NULL) {
		_nss_dnsblock_syslog
		    ("ERROR: PCRE compilation failed at offset %d: %s\n",
		     erroffset, error);
		return 0;
	}

	rc = pcre_exec(re, NULL, subject, size, 0, 0, ovector, OVECCOUNT);

	if (rc < 0) {
		switch (rc) {
		case PCRE_ERROR_NOMATCH:
			break;
		default:
			/* Matching error */
			break;
		}
		pcre_free(re);
		return 0;
	}

	pcre_free(re);
	return 1;
}

void _nss_dnsblock_syslog(const char *format, ...)
{
	char log_buf[BUF_SIZE];
	char va_buf[BUF_SIZE];
	va_list vlist;

	memset(log_buf, '\0', BUF_SIZE);
	memset(va_buf, '\0', BUF_SIZE);

	va_start(vlist, format);
	vsnprintf(va_buf, BUF_OFF1, format, vlist);
	va_end(vlist);

	memcpy(log_buf, va_buf, BUF_OFF1);

	openlog(LOG_NAME, LOG_PID | LOG_CONS, LOG_USER);
	syslog(LOG_INFO, "%s", log_buf);
	closelog();
}

int _nss_dnsblock_valid_hostname(const char *hostname, int hostsize)
{
	int i = 0;

	for (i = 0; i < hostsize; i++) {
		if (hostname[i] >= '0' && hostname[i] <= '9') {
			continue;
		} else if (hostname[i] >= 'A' && hostname[i] <= 'Z') {
			continue;
		} else if (hostname[i] >= 'a' && hostname[i] <= 'z') {
			continue;
		} else if (hostname[i] == '-') {
			continue;
		} else if (hostname[i] == '_') {
			continue;
		} else if (hostname[i] == '.') {
			continue;
		} else {
			return 0;
		}
	}

	return 1;
}


src/main.h
/*
Copyright 2014 Aiko Barz

This file is part of dnsblock.

dnsblock is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

dnsblock is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with dnsblock.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "config.h"
#include "pcre.h"

#define OVECCOUNT 30

int _nss_dnsblock_lookup(const char *hostname, UCHAR * address,
			 int *af, int *address_size);
int _nss_dnsblock_load(const char *filename, const char *hostname,
		       UCHAR * address, int *af, int *address_size);
int _nss_dnsblock_pcre_match(const char *subject, const char *pattern);
void _nss_dnsblock_syslog(const char *format, ...);
int _nss_dnsblock_valid_hostname(const char *hostname, int hostsize);


src/config.h
/*
Copyright 2014 Aiko Barz

This file is part of dnsblock.

dnsblock is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

dnsblock is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with dnsblock.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef CONFIG_H
#define CONFIG_H

#define BUF_SIZE 1024
#define BUF_OFF1 1023

#define TRUE 1
#define FALSE 0

#define LOG_NAME "dnsblock"

typedef unsigned long int ULONG;
typedef unsigned short USHORT;
typedef unsigned int UINT;
typedef unsigned char UCHAR;
typedef long int LONG;

#ifdef IPV6
#define IP_SIZE_LIST 128
#define IP_SIZE_META_PAIR 18
#define IP_SIZE_META_PAIR8 144
#define IP_SIZE_META_TRIPLE 38
#define IP_SIZE_META_TRIPLE8 304
#define IP_ADDRLEN INET6_ADDRSTRLEN
typedef struct sockaddr_in6 IP;
typedef struct in6_addr IN_ADDR;
#elif IPV4
#define IP_SIZE_LIST 32
#define IP_SIZE_META_PAIR 6
#define IP_SIZE_META_PAIR8 48
#define IP_SIZE_META_TRIPLE 26
#define IP_SIZE_META_TRIPLE8 208
#define IP_ADDRLEN INET_ADDRSTRLEN
typedef struct sockaddr_in IP;
typedef struct in_addr IN_ADDR;
#endif

#endif				/* CONFIG_H */
